iocore/net/P_SSLNetVConnection.h

```diff
@@ -466,6 +490,8 @@ class SSLNetVConnection : public UnixNetVConnection, public ALPNSupport
   in_port_t tunnel_port       = 0;
   bool tunnel_decrypt         = false;
   X509_STORE_CTX *verify_cert = nullptr;
+
+  std::shared_ptr<std::string> _serverName;
```

Comment:
I thought we talked about storing this in the session and not the VC?  The VC from what I know sometimes won't be around when we do logging.

Reply:
The HttpSM object for each transaction has a shared pointer to the string object, so that its lifetime is extended till the last HttpSM object referring to it is destroyed.  If I put it in the session object (I guess you mean in the base class?), what mechanism or strategy ensures it lives longer than the VConn object?

Reply:
If you look at line 90 of the assembler, where the default-constructed shared pointer is initialized, it's just all zero bytes:  https://godbolt.org/z/ZytAmB .  So it seems safe to put in a class that's allocated from a freelist.

Reply:
@shinrich  is sure that, as long as an HttpSM object is alive, the session object of the session the HttpSM objects of the transactions that are/were part of will be alive, so it would be feasible to change this to put the server name in the session object.

Reply:
@shinrich @zwoop @bryancall I created a test version of TS with the commit https://github.com/ywkaras/trafficserver/commit/c74cf7bd97fec99b59c54702c4d9f712c6f7ace0 to verify that a ProxySession object is never destroyed before destroying the HttpSM objects for the transactions that are part of the session.  I ran all of the existing Au tests with this test version of TS.  It looks like the session object can be destroyed first:
```
wkaras ~/REPOS/TS/tests/_sandbox
M$ fgrep '.notified()' $(find $(find . -name \*-ts ) -name stream.stderr.txt )
./tls_session_cache/_output/tls_session_cache-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./tls_tunnel/_output/tls_tunnel-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./tls_tunnel/_output/tls_tunnel-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./tls_tunnel/_output/tls_tunnel-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./double_h2/_output/double_h2-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./double_h2/_output/double_h2-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./double_h2/_output/double_h2-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./slice/_output/slice-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./slice_error/_output/slice_error-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./tls_ticket/_output/tls_ticket-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./http408/_output/http408-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./via/_output/via-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./tls_tunnel_forward/_output/tls_tunnel_forward-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./double/_output/double-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./double/_output/double-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./double/_output/double-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./double/_output/double-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./syntax/_output/syntax-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./tls_forward_nonhttp/_output/tls_forward_nonhttp-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./setstatus/_output/setstatus-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
./deny0/_output/deny0-ts/stream.stderr.txt:Fatal: HttpSM.cc:342: failed assertion `!watcher.notified()`
wkaras ~/REPOS/TS/tests/_sandbox
M$ 
```
Therefore, I don't think it would work to store the client SNI server name in the session object.

Reply:
Unfortunately, I think Walt has today's most pragmatic solution. Yes ideally SNI would be stored on the ProxySession or in HostDBRecord. But until we get this code cleaned up this might have to do.

Reply:
Will take a look at walt's watcher.  Would be interesting to see what these cases are.  I spent a lot of effort to ensure that the session was still around while the HttpSM transaction hooks went off, so the session close hook could go last.

After that I am less sure the order in which the memory objects are freed.  After all the major work is done (and logging in this case), the order of the final destroys doesn't really matter.

Reply:
I need to make watcher 2.0, the way it is currently, it's possible the notified flag is being set in the HttpSM object after it's been reallocated for a new transaction in a new session.

Reply:
I ran with the watcher patch and could not see the problem.  But examining the log, I do see an option where for Http1 at least the session could cleanup before logging occurs (and before HttpSM finally is destroyed). 

I pushed a branch that adjusts the order of logging in HttpSM::kill_this that should avoid that path.  @ywkaras could you give this a try in your testing environment?

https://github.com/shinrich/trafficserver/tree/object-lifetime-vs-logging

Reply:
I changed the test code to make it more robust:  https://github.com/ywkaras/trafficserver/commit/e6cde1b1842f6bcdd153763de87638f851c56cc1

But I'm still seeing the session object be destroyed before the HttpSM object in some Au test cases:
```
wkaras ~/REPOS/TS/tests
M$ fgrep 'Died()' $(find $(find . -name \*-ts ) -name stream.stderr.txt )
./_sandbox/tls_session_cache/_output/tls_session_cache-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls_tunnel/_output/tls_tunnel-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls_tunnel/_output/tls_tunnel-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls_tunnel/_output/tls_tunnel-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/double_h2/_output/double_h2-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/slice/_output/slice-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/slice_error/_output/slice_error-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/remap_ws/_output/remap_ws-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls_ticket/_output/tls_ticket-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/http408/_output/http408-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls/_output/tls-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/via/_output/via-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls_tunnel_forward/_output/tls_tunnel_forward-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/double/_output/double-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/cache_and_req_body/_output/cache_and_req_body-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/syntax/_output/syntax-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/tls_forward_nonhttp/_output/tls_forward_nonhttp-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/setstatus/_output/setstatus-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
./_sandbox/deny0/_output/deny0-ts/stream.stderr.txt:Fatal: HttpSM.cc:343: failed assertion `!watcher.watchedDied()`
wkaras ~/REPOS/TS/tests
M$ 
```

Reply:
Yes, when I run this change with my improved test code, it does seem to  cause the Session object (and I would guess the ProxyTransaction object as well) to live as long as the HttpSM object.